jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1866,7 +1866,7 @@ private static String getNonArrayClassPackageName(Class<?> clz) {
 	String name = clz.getName();
 	int index = name.lastIndexOf('.');
 	if (index >= 0) {
-		return name.substring(0, index);
+		return name.substring(0, index).intern();
```

Comment:
The substring call here will create a new String (all be it short lived) which will then be passed to the interning logic. In Java 8 the backing array will be shared between the substring and the new String, but in Java 9 and later the substring will have to copy the backing array as well as create the new String object.

The intern will reduce the number of escaping strings so it isn't strictly worse. The unfortunate thing is that the intern operation will have to compute the string hashcode on the substring every time since substring doesn't derive it from the base string.

We also have to repeat the lastIndexOf and the clz.getName every time which is also computationally expensive.

Reply:
If we need the space for perf could one of the other fields (like the EnumVar field or methodHandleCache) move down the hierarchy to free up space for the package name since that is more generic?

Reply:
First to clarify that intention of this PR is to match RI behaviour discovered during another problem investigation. There is no performance testcase related to this change.
`Liberty DayTrader3 footprint` benchmark tests shows this change has little impact (mainly in noise level).

Agreed that computing expensive methods are invoked before `intern()`. Though it is can be avoided via a local cached instance field (or mechanism alike), will not implement it now until there is a use case showing measurable performance gain.

Reply:
Well we should fix the problem which this does - I think the perf improvement is worth doing as a follow-on if you expect this to be called more since we know it is an issue and it would be better to fix it before it becomes a performance problem

Reply:
It is not expected that `j.l.Class.getPackageName()` is called often enough to have a (throughput) performance impact. There is no testcase to support that. This PR avoids creating a new String object for each `j.l.Class.getPackageName()` call for all classes within same package.

Implementing local cache is going to introduce overhead such as an extra field (or alike) which is not desirable if there is no application demonstrating performance gain.